staging/vhost-device-gpu/src/virt_gpu.rs

```diff
@@ -393,6 +394,58 @@ impl VirtioGpu {
         Ok(OkNoData)
     }
 
+    /// Updates the cursor's memory to the given resource_id, and sets its position to the given
+    /// coordinates.
+    pub fn update_cursor(
+        &mut self,
+        _resource_id: u32,
+        gpu_backend: &mut GpuBackend,
+        cursor_pos: VhostUserGpuCursorPos,
+        hot_x: u32,
+        hot_y: u32,
+        _mem: &GuestMemoryMmap,
+    ) -> VirtioGpuResult {
+        let size: usize = 4096;
+        let data = vec![0u32; size];
+        let mut data_array = [0u32; 4096];
+        data_array.copy_from_slice(&data[..]);
+        let cursor_update: VhostUserGpuCursorUpdate = VhostUserGpuCursorUpdate {
+            pos: cursor_pos,
+            hot_x: hot_x,
+            hot_y: hot_y,
+            data: data_array,
+        };
+        gpu_backend.cursor_update(&cursor_update).map_err(|e| {
+            error!("Failed to update cursor pos from frontend: {}", e);
+            ErrUnspec
+        })?;
+        //todo: flush resource
+        //self.flush_resource(resource_id, gpu_backend, mem)
+        Ok(OkNoData)
+    }
+
+    /// Moves the cursor's position to the given coordinates.
+    pub fn move_cursor(
+        &mut self,
+        resource_id: u32,
+        gpu_backend: &mut GpuBackend,
+        cursor: VhostUserGpuCursorPos,
+    ) -> VirtioGpuResult {
+        if resource_id == 0 {
+            gpu_backend.cursor_pos_hide(&cursor).map_err(|e| {
+                error!("Failed to set cursor pos from frontend: {}", e);
+                ErrUnspec
+            })?;
+        } else {
+            gpu_backend.cursor_pos(&cursor).map_err(|e| {
+                error!("Failed to set cursor pos from frontend: {}", e);
+                ErrUnspec
+            })?;
+        }
+
+        Ok(OkNoData)
+    }
```

Comment:
This doesn't seem to follow the spec: 
> **VIRTIO_GPU_CMD_MOVE_CURSOR** Move cursor. Request data is struct virtio_gpu_update_cursor. Re-
sponse type is VIRTIO_GPU_RESP_OK_NODATA.
Move cursor to the place specified in pos. The other fields are not used and will be ignored by the
device.

Specifically the "The other fields are not used and will be ignored by the device." is not followed.  I don't think you should check the `resource_id` here and call `cursor_pos_hide`, this should probably only call `cursor_pos` .

Reply:
this is a non-normative statement. the spec also says here that `Cursor will be loaded from the specified resource_id and will be moved to pos.` also they did not explicitly state what other fields will be ignored, becuase they could also be referring to the padding. Also in qemu depending if there is resource_id it does the following https://gitlab.com/qemu-project/qemu/-/blob/master/contrib/vhost-user-gpu/vhost-user-gpu.c#L969

Reply:
The sentence "Cursor will be loaded from the specified resource_id and will be moved to pos. " in spec is for `UPDATE_CURSOR` not `MOVE_CURSOR `.  

The spec is quite unclear, but if QEMU does this, then it should be fine